rts/motoko-rts/src/types.rs

```diff
@@ -256,63 +257,120 @@ impl Value {
     }
 
     /// Assumes that the value is a pointer and returns the pointer value. In debug mode panics if
-    /// the value is not a pointer.
+    /// the value is not a pointer or if it points to a stale object that has been forwarded.
     pub fn get_ptr(self) -> usize {
         debug_assert!(self.get().is_ptr());
         unskew(self.0 as usize)
     }
 
-    /// Get the object tag. In debug mode panics if the value is not a pointer.
+    /// Check that the forwarding pointer is valid.
+    #[inline]
+    pub unsafe fn check_forwarding_pointer(self) {
+        debug_assert!(
+            self.forward().get_ptr() == self.get_ptr()
+                || self.forward().forward().get_ptr() == self.forward().get_ptr()
+        );
+    }
+
+    /// Check whether the object's forwarding pointer refers to a different location.
+    pub unsafe fn is_forwarded(self) -> bool {
+        self.check_forwarding_pointer();
+        self.forward().get_ptr() != self.get_ptr()
+    }
+
+    /// Get the object tag. No forwarding. Can be applied to any block, regular objects
+    /// with a header as well as `OneWordFiller`, `FwdPtr`, and `FreeSpace`.
+    /// In debug mode panics if the value is not a pointer.
     pub unsafe fn tag(self) -> Tag {
-        debug_assert!(self.get().is_ptr());
-        (self.get_ptr() as *mut Obj).tag()
+        *(self.get_ptr() as *const Tag)
```

Comment:
I notice that many old calls to `.tag()` have been replaced by `.tag`. Is that just more efficient or some other reason?

Reply:
Thanks for asking this. The changes from `.tag()` to `.tag` are actually a remainder from earlier changes, where I added a forwarding check inside the `tag()` function. This was not compatible with the threading technique used in the compacting and generational GC. This is why I changed the code there to the low-level accesses. I will revert those superfluous changes. 

Actually, with the introduction of the mark bitmap, we could again consider swapping forwarding pointer and tag but it could be fragile if we still would like to perform a linear heap scan at some time... Maybe I can try in a separate PR to see how big the performance benefit would be.